Fix, send OA alerts after MP3 file processing is completed.

albertisfu opened the pull request:
This PR fixes the issue identified with sending/scheduling alerts that lack a `local_path` file and a `duration` in the sent alert. The problem arose because alerts were being sent/scheduled before the MP3 file processed by `process_audio_file` was ready.

To address this, the `abort_es_audio_indexing` check has been introduced into the `save_document_in_es` task. This abort OA indexing/percolator actions if `process_audio_file` has not been completed ('processing_complete' is not received in the update_fields save() parameters).

This check allows OA document updates without requiring 'processing_complete' in `update_fields` when the document already exists in ES.

Consequently, only OA documents whose MP3 files have been processed will be indexed for first time into ES and will trigger/schedule alerts.

**Regarding the other issue:**

> View Full Results isn't ordered by "latest first"

I verified that Solr alerts are ordered by `score desc`. Can we confirm that alerts should now be ordered by "Latest First" (sorted by dateArgued desc)?

O also notice the following difference:

When clicking "View Full Results" in Solr, RT Alerts show results with a `dateArgued` after 10 days ago. Daily alerts show results whose `dateArgued` matches the current day. Weekly alerts show results with a `dateArgued` after 7 days ago. Monthly alerts show results with a `dateArgued` after 28 days ago.

Currently, ES OA RT Alerts contain only one hit, since each document is percolated and triggers alerts independently. My question is: for RT Alerts, when the user clicks "View Full Results" should we display all results matching the query or should we apply a date-range filter? If so, what should the time frame be?

For daily, weekly, and monthly alerts, when the user clicks "View Full Results" I assume we should continue showing results within date ranges that match the current day, the past 7 days, and the past 28 days, respectively, correct?

There are some failing tests due to a `ReadTimeout`. I'm also experiencing this issue locally in this branch and also in main. 
It seems related to a github timeout when running `harvard_opinions` tests, maybe this is something temporarily.

`HTTPSConnectionPool(host='raw.githubusercontent.com', port=443)`

mlissner commented on cl/scrapers/management/commands/cl_scrape_oral_arguments.py:
```diff
@@ -173,7 +173,8 @@ def scrape_court(
                 process_audio_file.apply_async(
-                    (audio_file.pk,), countdown=random.randint(0, 3600)
+                    args=(audio_file.pk,),
+                    countdown=random.randint(0, 3600),
```
I'm glad you tweaked the formatting here, so I could see this. The jitter we put here was to help with our old architecture so that a bunch of new audio files wouldn't clog up the server. We're more scalable now, so I'd say we should take the random jitter out of here and we should be fine. 

Otherwise, some RT alerts will be delayed by an hour, which isn't great.

albertisfu replied:
Great, I've removed the jitter from this task so now is just: `process_audio_file.delay(audio_file.pk)`

mlissner commented on cl/search/tasks.py:
```diff
@@ -172,22 +173,64 @@ def delete_items(items, app_label, force_commit=False):
     :return: SaveDocumentResponseType or None
     """
 
+    if abort_es_audio_indexing(instance, es_document, update_fields):
+        # This check is required to avoid indexing and triggering search alerts
+        # for Audio instances whose MP3 files have not yet been processed by
+        # process_audio_file.
+        self.request.chain = None
+        return None
```
This looks fine, but it'd be nicer if we can avoid creating the chain of tasks in the first place. Can we move this check upstream to the places where we create this task?

albertisfu replied:
Sure, I've moved this check upstream into the signal handler, so now the chain is avoided at all.

mlissner reviewed (changes requested):
A couple of quick thoughts. Thank you!

albertisfu commented:
Thanks, I've applied the requested changes here and also implemented the modifications we discussed in the meeting.

**Tweaks to the View Full Results query in alerts:**

- I applied the `dateArgued desc` ordering for OA and `dateFiled desc` for other search types.
- I also added a date range filter: `argued_after` for OA and `filed_after` for other search types. For scheduled alerts, a cut-off date is passed to the query based on the rate: daily uses the current day, weekly uses 7 days ago, and monthly uses the first day of the previous month. This is how it worked in Solr.
- For RT Alerts, we are not adding a range filter for now. The `View Full Results` query will show all results matched by the query. Later, we'll incorporate the [date_created](https://github.com/freelawproject/courtlistener/issues/3109) field for filtering hits around the time the alert was triggered.

I noticed that email text alerts don't have a `View Full Results` link. Instead, they only include:
`Edit / Delete this Alert: https://www.courtlistener.com/alert/edit/11/`
Is there a reason it's not the same as in the HTML version? Should we use the same URL as in the HTML, which includes the alert query?

**Also, regarding `View Full Results`**, I noticed that in Solr, the hits in an alert are limited to 20.

Currently, for scheduled alerts, we don't have this limit; we store all the hits triggered for a user and alert. So, when sending daily, weekly, or monthly alerts, we include all the hits per alert grouped in the email.

Should we impose a limit of 20 hits on scheduled alerts? I believe the best approach is to store only up to 20 hits per alert, user, and rate, rather than storing everything and adding the limit during grouping.

mlissner commented:
> I noticed that email text alerts don't have a `View Full Results` link. Instead, they only include: `Edit / Delete this Alert: https://www.courtlistener.com/alert/edit/11/` Is there a reason it's not the same as in the HTML version? Should we use the same URL as in the HTML, which includes the alert query?

No idea why it's not the same. It should be. Let's do that.

> **Also, regarding `View Full Results`**, I noticed that in Solr, the hits in an alert are limited to 20.
> 
> Currently, for scheduled alerts, we don't have this limit; we store all the hits triggered for a user and alert. So, when sending daily, weekly, or monthly alerts, we include all the hits per alert grouped in the email.
> 
> Should we impose a limit of 20 hits on scheduled alerts? I believe the best approach is to store only up to 20 hits per alert, user, and rate, rather than storing everything and adding the limit during grouping. The downside of this approach is that we won't know the total number of hits for the alert. We could show something like: `Your daily oral argument alert — Courts: All — had 20+ hits:`

This approach makes sense to me, but we can go higher than 20. 30, say?

mlissner reviewed (commented):
Looks good to me. Just need the couple of tweaks you explained in your comment. Rest is ready.

albertisfu commented:
- Changes are applied, now the text version of the alert contains the correct link for the `View Full Results/Edit` button.

- Also now we only store the number on hits set in the setting `SCHEDULED_ALERT_HITS_LIMIT`, set to `30`.

So Alerts look like these (tests with hit limit of 3):

```
*****************
CourtListener.com
*****************

-------------------------------------------------------
We have news regarding your alerts at CourtListener.com
-------------------------------------------------------

Your daily oral argument alert -- Test Alert OA Daily 1 -- had 3+ hits:
-------------------------------------------------------
View Full Results / Edit this Alert: https://www.courtlistener.com/?q=USA+vs+Bank+&type=oa&order_by=dateArgued+desc&argued_after=09%2F08%2F2023&edit_alert=177/
Disable this Alert (one click): https://www.courtlistener.com/alert/disable/IYdoQMyzHt1don12j1ceLYS1USZf4Rz8etCeUZWu/

1. USA vs Bank (Bankr. C.D. Cal. 2023)
Date Argued: September 8th, 2023 | Docket Number: 19-5735 | Duration: 1

~~~~~
 - View this item on our site: https://www.courtlistener.com/audio/137/usa-vs-bank/
 - Download original from the court: https://johnson.com/
 - Download the original from our backup: https://storage.courtlistener.com/dat/2023/09/08/example.dat

2. USA vs Bank (Bankr. C.D. Cal. 2023)
Date Argued: September 8th, 2023 | Docket Number: 19-5735 | Duration: 1

~~~~~
 - View this item on our site: https://www.courtlistener.com/audio/139/usa-vs-bank/
 - Download original from the court: http://www.mcbride.com/
 - Download the original from our backup: https://storage.courtlistener.com/dat/2023/09/08/example_EK4oYq9.dat

3. USA vs Bank (Bankr. C.D. Cal. 2023)
Date Argued: September 8th, 2023 | Docket Number: 19-5735 | Duration: 1
```

The Alert hit limit is set by each alert in the email independently:

![Screenshot 2023-09-08 at 12 21 28](https://github.com/freelawproject/courtlistener/assets/486004/e8a55816-7037-42f5-a31c-07c2014a812b)

mlissner commented:
Shouldn't the `+` only be there if there are more than 30 hits?

albertisfu commented:
Yeah, well it's not possible to know if there are more than 30 hits in this approach as I mentioned above:

> The downside of this approach is that we won't know the total number of hits for the alert. We could show something like: Your daily oral argument alert — Courts: All — had 20+ hits

But we can only show the "+" if the number of results has reached the `SCHEDULED_ALERT_HITS_LIMIT`.

I applied the fix, so now if there are less than the limit set, the "+" is not displayed.

![Screenshot 2023-09-08 at 16 21 51](https://github.com/freelawproject/courtlistener/assets/486004/70004c52-7055-4645-a4ae-e6bd04b59d85)

Sometimes the real number of results can be equal to `SCHEDULED_ALERT_HITS_LIMIT` and "+" will be displayed even there are no more than `30`.

Is that ok?

mlissner commented:
Yeah, I figured we'd wind up with these trade offs. One thing: Your screenshot shoes 3+. I'm guessing that's because you had it configured for 3, not 30?

albertisfu commented:
Yeah that's correct, the screenshoot shows `3+` due to I set `SCHEDULED_ALERT_HITS_LIMIT` to `3` in my settings. In prod it'll be `30`.

mlissner reviewed (approved):
Changes look good. I'll review the comments and I think we're good to go here.

mlissner commented:
Automerge enabled.
